File: 515. Find Largest Value in Each Tree Row/solution.py
#using BFS
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def largestValues(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        queue = deque([root])
        res = []
        while queue:
            n = len(queue)
            max_node = -98765432199877
            for i in range(n):
                node = queue.popleft()
                if node:
                    max_node = max(node.val,max_node)
                if node and node.left:
                    queue.append(node.left)
                if node and node.right:
                    queue.append(node.right)
            res.append(max_node)            
        return res 
        
#using DFS
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def largestValues(self, root: Optional[TreeNode]) -> List[int]:
        def find(root,res,level):
            if not root:
                return []
            if level == len(res):
                res.append(root.val)    
            else:
                res[level] = max(res[level],root.val)
            find(root.left,res,level+1)
            find(root.right,res,level+1)
        res = []  
        find(root,res,0)  
        return res  
                

File: 515. Find Largest Value in Each Tree Row/solution.rs
//using BFS
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::collections::VecDeque;
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn largest_values(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut q:VecDeque<Rc<RefCell<TreeNode>>> = VecDeque::new();
        if let Some(first) = root.as_ref(){
            q.push_back(first.clone());
        }
        let mut res = Vec::new();
        while !q.is_empty(){
            let n = q.len();
            let mut max_val = i32::MIN;
            for i in 0..n{
                if let Some(n) = q.pop_front(){
                    
                        let node = n.borrow();
                        max_val = std::cmp::max(node.val,max_val);
                        if let Some(left) = node.left.as_ref(){
                            q.push_back(left.clone());
                        }
                        if let Some(right) = node.right.as_ref(){
                            q.push_back(right.clone());
                        }
                    
                }
            }res.push(max_val);
        } res

    }
}
//using DFS
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn dfs(root: Option<Rc<RefCell<TreeNode>>>, res:&mut Vec<i32>,level:usize){
        if root.as_ref() == None{
            return 
        }
       
            if let Some(n) = root{
            let node = n.borrow();
            if level == res.len(){
             res.push(node.val);
            }
            else{
                res[level] = std::cmp::max(node.val,res[level]);
            }
            Self::dfs(node.left.clone(),res,level+1);
            Self::dfs(node.right.clone(),res,level+1);
            
        }
    }
    pub fn largest_values(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut level = 0;
        let mut res = Vec::new();
        Self::dfs(root,&mut res,level);
        res
        
    }
}